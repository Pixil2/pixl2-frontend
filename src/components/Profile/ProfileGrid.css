.ProfileGrid {
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  grid-template-rows: repeat(10, 1fr);
  background-color: pink;
  height: 200px;
  width: 200px;
}
