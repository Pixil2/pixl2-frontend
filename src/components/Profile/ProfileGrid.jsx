import React, { useEffect, useState } from 'react';
import styles from './ProfileGrid.css';
import Pixl from '../Canvas/Pixl';
import { v4 as uuid } from 'uuid';

export default function ProfileGrid({ image, setImage, tool, eraser, color }) {
  const [pixelArray, setPixelArray] = useState([]);
  const [rainbowIndex, setRainbowIndex] = useState(0);
  const rainbowArray = [
    '#f54242',
    '#f59642',
    '#f5e942',
    '#84f542',
    '#42ddf5',
    '#b15beb',
    '#f779b4',
  ];

  function renderImage(imageObject) {
    const colorArray = imageObject.colorArray;
    const area = imageObject.height * imageObject.width;
    const image = [];

    for (let i = 0; i < area; i++) {
      if (imageObject.height >= imageObject.width) {
        image.push(
          <Pixl
            pixelIndex={i}
            key={uuid()}
            background={colorArray[i]}
            width={`${Math.round(250 / imageObject.height)}px`}
            height={`${Math.round(250 / imageObject.height)}px`}
            handleClick={handleClick}
            index={i}
          />
        );
      } else {
        image.push(
          <Pixl
            pixelIndex={i}
            key={uuid()}
            background={colorArray[i]}
            width={`${Math.round(200 / imageObject.width)}px`}
            height={`${Math.round(200 / imageObject.width)}px`}
            handleClick={handleClick}
            index={i}
          />
        );
      }
    }

    setPixelArray(image);
  }

  const handleClick = (index) => {};

  useEffect(() => {
    renderImage(image);
  }, [image, tool, color]);

  return (
    <div
      className={styles.ProfileGrid}
      style={{
        gridTemplateColumns: `repeat(${image.width}, 1fr)`,
        gridTemplateRows: `repeat(${image.height}, 1fr)`,
      }}
    >
      {pixelArray.map((item) => item)}
    </div>
  );
}
