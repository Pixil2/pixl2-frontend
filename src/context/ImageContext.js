import { createContext, useEffect, useState } from 'react';
import { createImage } from '../utils/grid';

export const imageContext = createContext();

export function ProvideImage({ children }) {
  const [image, setImage] = useState({});
  const [eraser, setEraser] = useState({});
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    setImage(createImage('title', 15, 15));
    setEraser(createImage('eraser', 10, 10));
    setLoading(false);
  }, [loading]);

  return (
    <imageContext.Provider value={{ image, eraser, loading, setLoading }}>
      {children}
    </imageContext.Provider>
  );
}
