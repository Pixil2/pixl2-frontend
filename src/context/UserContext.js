import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { signIn, signOut } from '../services/users';

const UserContext = createContext();

export const UserProvider = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [user, setUser] = useState(null);

  const login = async (credentials) => {
    try {
      const user = await signIn(credentials);
      setUser(user);
    } catch (error) {
      throw error;
    }
  };

  const logout = useCallback(() => {
    signOut().then(() => setUser(null));
  }, []);
};
