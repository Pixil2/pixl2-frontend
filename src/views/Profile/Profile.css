.Profile {
  display: flex;
  gap: 6vw;
  justify-content: center;
}
