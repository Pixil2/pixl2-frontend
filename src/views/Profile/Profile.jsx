import { useEffect, useState } from 'react';
import { getUserImages, deleteImageById } from '../../services/images';
import ProfileGrid from '../../components/Profile/ProfileGrid';
import { getCurrentUser } from '../../services/users';
import { v4 as uuid } from 'uuid';
import styles from './Profile.css';
import { Link, useNavigate } from 'react-router-dom';
import Header from '../../components/Layout/Header';
import { getTagByImageId } from '../../services/tags';

export default function Profile() {
  const [currentImages, setCurrentImages] = useState([]);
  const navigate = useNavigate();

  useEffect(() => {
    const fetch = async () => {
      const user = await getCurrentUser();
      const res = await getUserImages(user.id);
      const tagArray = [];
      await Promise.all(
        res.map(async (item) => {
          const thing = await getTagByImageId(item.id);
          tagArray.push({ ...thing });
        })
      );
      setCurrentImages(tagArray);
    };
    fetch();
  }, []);

  const handleEdit = (id) => {
    navigate(`/canvas/edit/${id}`);
  };

  const handleDelete = async (id) => {
    await deleteImageById(id);
    const newCurrentImages = currentImages.filter((item) => item.id !== id);
    setCurrentImages(newCurrentImages);
  };

  return (
    <div className={styles.Profile}>
      <Header />
      <div className={styles.profileHeader}>
        <h1 className={styles.profileTitle}>User Gallery</h1>
        <p className={styles.profileCaption}>
          Hey, nice work. Check out all of your amazing artwork below or get
          back in there and make some more!
        </p>
        <Link to="/canvas">
          <button className={`${styles.createButton} ${styles.imageButton}`}>
            CREATE IMAGE
          </button>
        </Link>
      </div>
      <div className={styles.ProfileContainer}>
        {currentImages.map((item, index) => {
          const tag = item.tags[0].name;
          return (
            <div key={uuid()}>
              <div className={styles.imageHeader}>
                <h1 className={styles.imageTitle}>{item.title}</h1>
                <p className={styles.imageTag}>{tag}</p>
              </div>
              <ProfileGrid image={item} />
              <div className={styles.imageButtons}>
                <button
                  className={styles.imageButton}
                  value={item.id}
                  onClick={(e) => handleEdit(e.target.value)}
                >
                  EDIT
                </button>
                <button
                  className={styles.imageButton}
                  value={item.id}
                  onClick={(e) => handleDelete(e.target.value)}
                >
                  DELETE
                </button>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
}
