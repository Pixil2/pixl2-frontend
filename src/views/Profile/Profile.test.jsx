import dotenv from 'dotenv';
dotenv.config();
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { UserProvider } from '../../context/UserContext';
import Profile from './Profile';

const image = {
  id: '1',
  title: 'A title',
  height: 10,
  width: 10,
  colorArray: ['rgb(0, 0, 0)'],
  userId: '100',
  isPublic: false,
  isApproved: null,
  tags: [{ name: 'nick cage' }],
};

const user = {
  id: '100',
  username: 'testUser',
};

const server = setupServer(
  rest.get(`${process.env.API_URL}/api/v1/images/user/100`, (req, res, ctx) => {
    return res(ctx.json([image]));
  }),
  rest.get(`${process.env.API_URL}/api/v1/images/1/tags`, (req, res, ctx) => {
    return res(ctx.json([image]));
  }),
  rest.get(`${process.env.API_URL}/api/v1/images`, (req, res, ctx) => {
    return res(ctx.json([image]));
  }),
  rest.get(`${process.env.API_URL}/api/v1/users/me`, (req, res, ctx) => {
    return res(ctx.json(user));
  }),
  rest.post(`${process.env.API_URL}/api/v1/users/me`, (req, res, ctx) => {
    return res(ctx.json(user));
  })
);
describe('Profile', () => {
  beforeAll(() => {
    // jest.useFakeTimers();
    server.listen();
  });

  afterAll(() => {
    server.close();
  });

  it.skip('Tests functionality of profile view', async () => {
    const { container } = render(
      <UserProvider>
        <MemoryRouter initialEntries={['/profile']}>
          <Routes>
            <Route path="/profile" element={<Profile />} />
          </Routes>
        </MemoryRouter>
      </UserProvider>
    );
    const loading = screen.getByText(/loading/i);
    expect(loading).toBeInTheDocument();
    await waitForElementToBeRemoved(() => screen.getByText(/loading/i));

    const createButton = await screen.findByRole('button', {
      name: /create image/i,
    });

    expect(createButton).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
