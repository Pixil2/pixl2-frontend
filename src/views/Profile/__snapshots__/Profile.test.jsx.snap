// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests functionality of profile view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Profile"
      >
        <header
          class="Header"
        >
          <h1
            class="headerLogo"
          >
            PIXL 2
          </h1>
          <div
            class="headerLinkContainer"
          >
            <a
              class="headerLink"
              href="/community"
            >
              Community Gallery
            </a>
            <a
              class="headerLink"
              href="/about"
            >
              About
            </a>
          </div>
        </header>
        <a
          href="/canvas"
        >
          <button>
            Create Image
          </button>
        </a>
        <div
          class="ProfileContainer"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Profile"
    >
      <header
        class="Header"
      >
        <h1
          class="headerLogo"
        >
          PIXL 2
        </h1>
        <div
          class="headerLinkContainer"
        >
          <a
            class="headerLink"
            href="/community"
          >
            Community Gallery
          </a>
          <a
            class="headerLink"
            href="/about"
          >
            About
          </a>
        </div>
      </header>
      <a
        href="/canvas"
      >
        <button>
          Create Image
        </button>
      </a>
      <div
        class="ProfileContainer"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
